pages/quests/index.tsx

```diff
@@ -8,24 +8,115 @@ import {
   Input,
   Stack
 } from '@chakra-ui/react'
-import { useAccount, useConnect } from 'wagmi'
-import { useRouter } from 'next/router'
+import { useEffect, useState } from 'react'
+import { useAccount, useConnect, useNetwork } from 'wagmi'
 import Link from 'next/link'
 import { useTranslation } from 'next-i18next'
 import { serverSideTranslations } from 'next-i18next/serverSideTranslations'
 import { Layout } from '@/components/layouts/layout'
 import { useMounted } from '@/hooks/useMounted'
 import { useKeywordSubmit } from '@/hooks/quest/useKeywordSubmit'
+import { useUpdateToken } from '@/hooks/quest/useUpdateToken'
 import { useHasNFT } from '@/hooks/useHasNFT'
+import { useBalanceOf } from '@/hooks/useBalanceOf'
+import { useTokenIdOf } from '@/hooks/useTokenIdOf'
+import { useTokenURI } from '@/hooks/useTokenURI'
+import { getContractAddress } from '@/utils/contractAddress'
+import { useIpfsSubmit, KamonToken } from '@/hooks/useIpfsSubmit'
+import { useToast } from '@/hooks/useToast'
+import { useIpfsGet } from '@/hooks/useIpfsGet'
 
 const Quests: NextPage = () => {
   const { t } = useTranslation('common')
+  const { toast } = useToast()
   const mounted = useMounted()
   const { connect, connectors } = useConnect()
   const [metaMask] = connectors
   const { data } = useAccount()
-  const { keyword, inputChange, submit, isSubmitting } = useKeywordSubmit()
+  const { keyword, inputChange, submit, isSubmitting, keywordSubmitSucceeded } = useKeywordSubmit()
   const { hasNFT } = useHasNFT()
+  const { activeChain } = useNetwork()
+  const kamonNFT = getContractAddress({
+    name: 'kamonNFT',
+    chainId: activeChain?.id
+  })
+  const { balanceOf } = useBalanceOf(kamonNFT, data?.address)
+  const { tokenIdOf } = useTokenIdOf(kamonNFT, data?.address)
+  const { tokenURI } = useTokenURI(kamonNFT, tokenIdOf?.toNumber() || 0)
+  const [tokenId, setTokenId] = useState<BigInt>(BigInt(0))
+  const [tokenJSON, setTokenJSON] = useState<KamonToken>()
+  const [finalTokenUri, setFinalTokenUri] = useState('')
+  const [newTokenImageURI, setNewTokenImageURI] = useState('')
+  const [newTokenJSON, setNewTokenJSON] = useState<KamonToken>()
+  const [updateTxFailed, setUpdateTxFailed] = useState<boolean>()
+
+  const { ipfsSubmit, ipfsSubmitSucceeded, ipfsSubmitIsSubmitting } = useIpfsSubmit()
+  const { ipfsGet, ipfsGetIsSubmitting } = useIpfsGet()
+  const { updateToken, updateTokenSucceeded, updateTokenIsSubmitting } = useUpdateToken(
+    kamonNFT,
+    tokenId,
+    finalTokenUri,
+  )
+  const submittingSomething = () => {
+    return isSubmitting || ipfsSubmitIsSubmitting || ipfsGetIsSubmitting || updateTokenIsSubmitting
+  }
+  useEffect(() => {
+    if (balanceOf && tokenIdOf && tokenURI) {
```

Comment:
do you need all of the condition ?

Reply:
I have tried many things, and after taking some out I think maybe I do.

Note that some very obvious ones are there for TypeScript to believe that the variables will behave as expected.